Added autosave on the filters page

thegitone23 opened the pull request:
<!-- Please replace #XX below with an existing issue number. Remove the line entirely if none exist. -->
Fixes #41 

**Description**
This PR aims to add auto save feature on the filters section of the options page

**Other information**
<!-- Add any other information below or delete the "Other Information" line entirely. -->
An issue currently exists and i would find it very helpful to get some suggestions on fixing the issue, chrome gave me the following error,
![Capture](https://user-images.githubusercontent.com/33160019/68479840-53d47980-0259-11ea-9c39-a0528aca54d2.PNG)


**Checklist:**
- [ ] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [ ] My pull request targets the `master` branch of the repository.
- [ ] My commit messages follow [best practices][best_practices].
- [ ] I added tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [ ] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

<!-- Make sure you read and understand the following attestation. -->

Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.

thegitone23 committed:
changed the input field to avoid overflow for large text input

thegitone23 committed:
fixed the search input overflow issue while keeping input box and button icons in the same flow

thegitone23 committed:
merged upstream changes

thegitone23 committed:
Merge branch 'master' of https://github.com/creativecommons/ccsearch-browser-extension

thegitone23 committed:
added autosave feature to the filters options page

makkoncept commented:
Hi @thegitone23, thanks for the PR. 
The error you mentioned is [described here](https://developer.chrome.com/apps/storage). The maximum number of the set, remove, or clear operations that can be performed each minute is set to 120. A `runtime.lastError` is triggered if this limit is exceeded. Because of this error, after a couple of changes in the filters section, the extension will not write those changes to the local storage. 
You should add a check for the `chrome.runtime.lastError` in the callback of `chrome.storage.sync.set` function [here](https://github.com/creativecommons/ccsearch-browser-extension/blob/master/src/options/helper.js#L58). 
Something like the following would work
```
() => {
   if (chrome.runtime.lastError) {
        showNotification("Too many changes! Please wait some time", "negative", "snackbar-options");
   } else {
        showNotification("Settings saved!", "positive", "snackbar-options");
  }
}
```

makkoncept reviewed (changes requested):
There are some linting errors that you need to be fixed before merging. 
I am tagging some of them, but you can run the command `npm run lint` to list all the errors.

makkoncept commented on src/options/helper.js:
```diff
@@ -24,6 +24,7 @@ function addProvidersToDom(providers) {
     input.type = 'checkbox';
     input.id = providers[key];
     input.classList = 'vocab choice-field magenta-colored small-sized';
+    input.addEventListener("change",delayedSaveFiltersOptions);
```
Use a single quote instead of double.
There should be a space after the comma.

makkoncept commented on src/options/helper.js:
```diff
@@ -63,6 +66,10 @@ export function saveSingleFilter(inputElements) {
 }
 
+export function delayedSaveFiltersOptions() {
+  setTimeout(saveFiltersOptions, 1500);
```
'saveFilterOptions' was used before it was defined

makkoncept commented on src/options/helper.js:
```diff
@@ -45,6 +46,8 @@ export async function init() {
   const providers = await getLatestProviders();
   addProvidersToDom(providers);
+  [...elements.useCaseInputs].map((x) => {x.addEventListener('change', delayedSaveFiltersOptions)});
+  [...elements.licenseInputs].map((x) => {x.addEventListener('change', delayedSaveFiltersOptions)});
```
'delayedSaveFiltersOptions' was used before it was defined.

thegitone23 committed:
fixed linting errors and added warning message if filter settings are being updated too often

thegitone23 commented:
> 
> 
> Hi @thegitone23, thanks for the PR.
> The error you mentioned is [described here](https://developer.chrome.com/apps/storage). The maximum number of the set, remove, or clear operations that can be performed each minute is set to 120. A `runtime.lastError` is triggered if this limit is exceeded. Because of this error, after a couple of changes in the filters section, the extension will not write those changes to the local storage.
> You should add a check for the `chrome.runtime.lastError` in the callback of `chrome.storage.sync.set` function [here](https://github.com/creativecommons/ccsearch-browser-extension/blob/master/src/options/helper.js#L58).
> Something like the following would work
> 
> ```
> () => {
>    if (chrome.runtime.lastError) {
>         showNotification("Too many changes! Please wait some time", "negative", "snackbar-options");
>    } else {
>         showNotification("Settings saved!", "positive", "snackbar-options");
>   }
> }
> ```

First of all thanks for the review, I fixed the linting errors and also added the message as you suggested, but as you saw the way I implemented auto save is essentially giving every checkbox save button like behavior which is a little  inefficient, can you kindly suggest some improvements over the current method.

makkoncept commented:
Hi, @thegitone23. Sorry for the late response, but after testing the changes locally for some days I think adding this feature to the extension will not be user-friendly. This is because of the `MAX_WRITE_OPERATIONS_PER_MINUTE` limit. Only after a couple of changes, the limit error is shown and this will become worse in the future as we might add more data in the localStorage. The users may get confused due to this. 
We will not be adding this feature(at least for now).
Thanks for helping us out with this.
